src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -242,42 +244,138 @@ class PlacementConstraint(NamedTuple('PlacementConstraint', [
     __slots__ = ()
 
     @staticmethod
-    def from_dict(constraint: dict) -> "PlacementConstraint":
+    def from_dict(masters: List[MacroSize], constraint: dict) -> "PlacementConstraint":
         constraint_type = PlacementConstraintType.from_str(
             str(constraint["type"]))
+
+        ### Margins ###
+        # This field is mandatory in TopLevel constraints
+        # This field is disallowed otherwise
         margins = None  # type: Optional[Margins]
-        orientation = None  # type: Optional[str]
-        top_layer = None  # type: Optional[str]
-        layers = None  # type: Optional[List[str]]
-        obs_types = None  # type: Optional[List[ObstructionType]]
-        if constraint_type == PlacementConstraintType.TopLevel:
+        if "margins" in constraint:
+            assert constraint_type == PlacementConstraintType.TopLevel, "Non-TopLevel constraint must not contain margins: {}".format(constraint)
             margins_dict = constraint["margins"]
             margins = Margins(
-                left=float(margins_dict["left"]),
-                bottom=float(margins_dict["bottom"]),
-                right=float(margins_dict["right"]),
-                top=float(margins_dict["top"])
+                left=Decimal(str(margins_dict["left"])),
+                bottom=Decimal(str(margins_dict["bottom"])),
+                right=Decimal(str(margins_dict["right"])),
+                top=Decimal(str(margins_dict["top"]))
             )
+        else:
+            assert constraint_type != PlacementConstraintType.TopLevel, "TopLevel constraint must contain margins: {}".format(constraint)
+
+        ### Orientation ###
+        # This field is disallowed in TopLevel constraints
+        # This field is optional otherwise
+        orientation = None  # type: Optional[str]
         if "orientation" in constraint:
+            assert constraint_type != PlacementConstraintType.TopLevel, "Non-TopLevel constraint must not contain orientation: {}".format(constraint)
             orientation = str(constraint["orientation"])
+
+        ### Top layer ###
+        # This field is optional in HardMacro constraints
+        # This field is disallowed otherwise
+        top_layer = None  # type: Optional[str]
         if "top_layer" in constraint:
+            assert constraint_type == PlacementConstraintType.HardMacro, "Non-HardMacro constraint must not contain top_layer: {}".format(constraint)
```

Comment:
I also use this with hierarchical blocks. There is something fancier you could do to auto use pin layers but I think we might still want this control sometimes.

Reply:
OK, sounds good. I'll add it back to hierarchical.